### PR TITLE
feat: [WD-13704] Create image from backup file

### PR DESCRIPTION
**Pull request by Kxiru**
## Done

- Added import tarball button as well as associated modal.
- Minor scss changes
- Additional API calls for importing the file and amending image properties.

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - Upload a unified tarball image or a split tarball image to LXD-UI by selecting the "Import from Tarball" button on the Image list page.

## Screenshots

image
image
image

### CODE REVIEW

File: src/api/images.tsx [hunk 1]
```diff
@@ -114,3 +114,79 @@ export const createImageAlias = (
       .catch(reject);
   });
 };
+
+export const createImageFromInstance = (
```
Comment (mas-who) [inline src/api/images.tsx:118]: I think we should combine these functions into a single function `createImage`. This means we should try make adjustment to the existing `createImageFromInstanceSnapshot` function as well. The construction of the payload body should be determined outside of the API layer. Also be careful and don't forget about the `?project=` query parameter.

File: src/api/images.tsx [hunk 2]
```diff
@@ -114,3 +114,79 @@ export const createImageAlias = (
...
+      .then(resolve)
+      .catch(reject);
+  });
+};
+
+export const createImageFromTarball = (
+  fileList: FileList,
+): Promise<LxdOperationResponse> => {
+  return new Promise((resolve, reject) => {
+    let obj: File | FormData | null = null;
```
Comment (mas-who) [inline src/api/images.tsx:144]: the logic here could be moved to a util file

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+            eventQueue.set(
+              operation.metadata.id,
+              (event) => {
+                const fingerprint = event.metadata.metadata?.fingerprint ?? "";
+                if (values.alias) {
+                  void createImageAlias(fingerprint, values.alias);
+                }
+                if (values.isPublic) {
+                  void changeImagePublicity(fingerprint, values.isPublic);
```
Comment (mas-who) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:72]: we do not need to update the image to make it public here. You can set the public field for an image with the `X-LXD-public` header, refer to the [doc](https://documentation.ubuntu.com/lxd/en/latest/api/#/images/images_post). This is of course for a file upload request to `POST /1.0/images`.

File: src/api/images.tsx [hunk 4]
```diff
@@ -114,3 +114,79 @@ export const createImageAlias = (
...
+      method: "POST",
+      body: obj,
+    })
+      .then(handleResponse)
+      .then(resolve)
+      .catch(reject);
+  });
+};
+
+export const changeImagePublicity = (
```
Comment (mas-who) [inline src/api/images.tsx:177]: this function is not needed. See later comment.

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+} from "@canonical/react-components";
+import * as Yup from "yup";
+import { Link, useNavigate } from "react-router-dom";
+
+interface Props {
+  close: () => void;
+  project: string;
+}
+
+const ImportImageFromTarballForm: FC<Props> = ({ close, project }) => {
```
Comment (mas-who) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:25]: we should implement progress status for file upload similar to how it's done in `UploadCustomIso`. @edlerd I see we are using `axios` for this, although is it possible to track multiple file uploads at the same time (for the split tarball import)?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    },
+    validationSchema: Yup.object().shape({
+      alias: Yup.string(),
+    }),
+    onSubmit: (values) => {
+      if (fileList) {
+        createImageFromTarball(fileList)
+          .then((operation) => {
+            toastNotify.info(<>Creation of image from file started.</>);
+            close();
```
Comment (mas-who) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:62]: If we are tracking the file upload progress, then we should not close the modal here

**Review by mas-who — COMMENTED** [review]
Still need to do QA and go through the notification copy. Left some code feedback so long :)

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+interface Props {
+  close: () => void;
+  project: string;
+}
+
+const ImportImageFromTarballForm: FC<Props> = ({ close, project }) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const navigate = useNavigate();
+  const [fileList, setFileList] = useState<FileList | null>(null);
```
Comment (mas-who) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:29]: I don't think we need this state, we can manage `fileList` using `formik`

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    const published = (
+      <Link to={`/ui/project/${project}/images`}>published</Link>
+    );
+    toastNotify.success(<>Image {published} from file.</>);
+  };
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      const files = e.target.files;
+      setFileList(files);
```
Comment (mas-who) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:41]: instead of managing file with additional state, you can use `formik.setFieldValue` here

**Comment by mas-who** [comment]
@Kxiru QA for the importing functionality looks good, I was able to:
- import container and VM split tarballs
- import container and VM unified tarballs

File: src/api/images.tsx (hunk 1 above)
Comment (Kxiru) [reply src/api/images.tsx:118]: Are you saying that createImageAlias should be integrated into both createImagefromInstanceSnapshot and createImageFromInstance?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx (hunk 5 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:25]: @edlerd , should we implement an axios approach for this?

Comment (edlerd) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:25]: I think axios is a nice helper to get the upload status. We can have a progress bar per file that is being uploaded, or restart the same progress bar for each new file.

File: src/api/images.tsx (hunk 1 above)
Comment (mas-who) [reply src/api/images.tsx:118]: no I meant both `createImageFromInstanceSnapshot` and `createImageFromInstance` should become `createImage` with a body input determined from outside of the function.

Comment (edlerd) [reply src/api/images.tsx:118]: I think the `createImageFromInstance` function shouldn't be in this PR, it is introduced in https://github.com/canonical/lxd-ui/pull/833
though I agree, we could have a single `createImage` function, which receives a body. Then we can reuse it for the tarball upload and for the creation of an image from an instance.

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx (hunk 8 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:41]: I actually did try this before this implementation and it did not work. I also read somewhere online that for files, it is unsupported. I recreated it again just in case I might have done it wrong, but I got the same error: 
image

For reference, this was the new implementation:
```
  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      formik.setFieldValue("fileList", e.target.files);
    }
  };
```

Comment (mas-who) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:41]: You can refer to #853 for how this can be done. The key thing here is to remove `formik.getFieldProps` from the file input and let `changeFile` handle selection changes

File: src/api/images.tsx (hunk 1 above)
Comment (Kxiru) [reply src/api/images.tsx:120]: Created an issue to address this: https://warthogs.atlassian.net/browse/WD-14347?atlOrigin=eyJpIjoiMjcyYmI5Mjk3MjI1NGYxNWI4NWRlZGEwODViYjhlNDUiLCJwIjoiaiJ9

Comment (edlerd) [reply src/api/images.tsx:120]: Thanks

File: src/pages/images/actions/ImportFromTarballBtn.tsx [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+  return (
+    <>
+      {isOpen && (
+        <Portal>
+          <ImportImageFromTarballForm close={closePortal} project={project} />
+        </Portal>
+      )}
+      <ActionButton className="u-no-margin--bottom" onClick={openPortal}>
+        <span>Import from Tarball</span>
```
Comment (edlerd) [inline src/pages/images/actions/ImportFromTarballBtn.tsx:21]:
```suggestion
        <span>Upload image</span>
```
Maybe clearer?
Also this component could be UploadImageBtn and the form UploadImageForm, wdyt?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+            eventQueue.set(
+              operation.metadata.id,
+              (event) => {
+                const fingerprint = event.metadata.metadata?.fingerprint ?? "";
+                if (values.alias) {
+                  void createImageAlias(fingerprint, values.alias);
+                }
+                close();
+                navigate(0);
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:73]: What is this here for?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+            toastNotify.info(<>Creation of image from file started.</>);
+
+            eventQueue.set(
+              operation.metadata.id,
+              (event) => {
+                const fingerprint = event.metadata.metadata?.fingerprint ?? "";
+                if (values.alias) {
+                  void createImageAlias(fingerprint, values.alias);
+                }
+                close();
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:72]: The close should also be called in the failure case. It might be simplest to move the close call to the 4th argument of eventqueue.set, which is a callback that is called in both cases, on success and failure.

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+const ImportImageFromTarballForm: FC<Props> = ({ close, project }) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const navigate = useNavigate();
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+
+  const notifySuccess = () => {
+    const imported = <Link to={`/ui/project/${project}/images`}>imported</Link>;
+    toastNotify.success(<>Image {imported} from file.</>);
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:33]:
```suggestion
    toastNotify.success(<>Image {imported}.</>);
```
More concise and just as precise?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+                if (values.alias) {
+                  void createImageAlias(fingerprint, values.alias);
+                }
+                close();
+                navigate(0);
+                notifySuccess();
+              },
+              (msg) => {
+                toastNotify.failure(
+                  `Image creation from file failed.`,
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:78]:
```suggestion
                  `Image upload failed.`,
```

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+                  new Error(msg),
+                );
+              },
+            );
+          })
+          .catch((e) => {
+            toastNotify.failure(`Image creation from file failed.`, e);
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:85]:
```suggestion
            toastNotify.failure(`Image upload failed.`, e);
```

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+const ImportImageFromTarballForm: FC<Props> = ({ close, project }) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const navigate = useNavigate();
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+
+  const notifySuccess = () => {
+    const imported = <Link to={`/ui/project/${project}/images`}>imported</Link>;
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:32]:
```suggestion
    const uploaded = <Link to={`/ui/project/${project}/images`}>uploaded</Link>;
```

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+            Cancel
+          </Button>
+          <ActionButton
+            appearance="positive"
+            className="u-no-margin--bottom"
+            loading={formik.isSubmitting}
+            disabled={!formik.isValid}
+            onClick={() => void formik.submitForm()}
+          >
+            Create image
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:125]:
```suggestion
            Upload image
```

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+          </ActionButton>
+        </>
+      }
+    >
+      <Form onSubmit={formik.handleSubmit}>
+        <Input
+          type="file"
+          name="fileList"
+          label="Local file"
```
Comment (edlerd) [inline src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:134]:
```suggestion
          label="Image backup file"
```

File: src/sass/_page_header.scss [hunk 18]
```diff
@@ -24,7 +24,6 @@
 
   .page-header__base-actions {
     align-self: flex-start;
-    flex: 1;
```
Comment (edlerd) [inline src/sass/_page_header.scss:27]: If we remove the flex, we can also remove the align self and the whole block, as it won't have any effect.

File: src/util/helpers.tsx [hunk 19]
```diff
@@ -290,3 +290,27 @@ export const getClientOS = (userAgent: string) => {
 
   return null;
 };
+
+export const getImageToTarballBody = (fileList: FileList): File | FormData => {
```
Comment (edlerd) [inline src/util/helpers.tsx:294]: This seems very specific to the ImportImageFromTarballForm, so I would keep it local to that file. This is very unlikely to be reused anywhere else.

File: src/util/helpers.tsx [hunk 20]
```diff
@@ -290,3 +290,27 @@ export const getClientOS = (userAgent: string) => {
...
+  let obj: File | FormData | null = null;
+  const formData = new FormData();
+
+  if (fileList.length === 1) {
+    obj = fileList[0];
```
Comment (edlerd) [inline src/util/helpers.tsx:299]:
```suggestion
    return fileList[0];
```

This way we don't need the else, and we don't need to define obj above. We can directly return formData below.

File: src/util/helpers.tsx [hunk 21]
```diff
@@ -290,3 +290,27 @@ export const getClientOS = (userAgent: string) => {
 
   return null;
 };
+
+export const getImageToTarballBody = (fileList: FileList): File | FormData => {
+  let obj: File | FormData | null = null;
+  const formData = new FormData();
```
Comment (edlerd) [inline src/util/helpers.tsx:296]: This should be defined closer to its use case.

File: src/api/images.tsx [hunk 22]
```diff
@@ -116,6 +118,7 @@ export const createImageAlias = (
 };
 
 export const createImageFromInstance = (
+  //Also known as create Image from an Instance.
```
Comment (edlerd) [inline src/api/images.tsx:121]:
```suggestion
```
Seems unrelated?

File: src/pages/images/actions/ImportFromTarballBtn.tsx (hunk 9 above)
Comment (Kxiru) [reply src/pages/images/actions/ImportFromTarballBtn.tsx:21]: Can do. I was following the naming convention of other buttons such as ExportImageTarballBtn, CreateInstanceFromImageBtn and UploadCustomIsoBtn. I can change this if you wish? I believe ImportFromTarballBtn is a bit more specific.

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx (hunk 10 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:73]: This refreshes the imageList page when the image is created as the created image does not display in the list unless it is rerendered/refreshed.

File: src/pages/images/actions/ImportFromTarballBtn.tsx (hunk 9 above)
Comment (edlerd) [reply src/pages/images/actions/ImportFromTarballBtn.tsx]: CreateInstanceFromImageBtn is overly specific, because there is also another way of creating instances without selecting an image first.

Here I think we only have one way to upload/create an image, that's why I suggested the simpler UploadImage name. Maybe also better to call it ExportImage, as there is only one way of export, so we don't need the Tarball in the name. Or do we have avariations of export / upload image? Thinking of it -- maybe it should be DownloadImage instead of ExportImage as that more close to what it does?

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx (hunk 10 above)
Comment (edlerd) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:93]: Should we better do a query cache invalidation for `[queryKeys.images, project]`? That should trigger a refresh of the cached images.

File: src/pages/images/actions/ImportFromTarballBtn.tsx (hunk 9 above)
Comment (Kxiru) [reply src/pages/images/actions/ImportFromTarballBtn.tsx]: Alrighty, coming right up!
ImportFromTarballBtn -> UploadImageBtn (and Form)
ExportImageTarballBtn -> DownloadImageBtn (and Form)

File: src/pages/images/actions/forms/ImportImageFromTarballForm.tsx (hunk 10 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/ImportImageFromTarballForm.tsx:93]: Perfect! That works great, thanks David :)

**Comment by Kxiru** [comment]
I believe this latest version is up to date, @edlerd , @mas-who . Please let me know your thoughts :)

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+        );
+      }
+
+      return formData;
+    }
+    return fileList[0];
+  };
+
+  const formik = useFormik<{
+    alias?: string;
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:65]:
```suggestion
    alias: string;
```
I think this is always string as it initializes with empty string.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+              operation.metadata.id,
+              (event) => {
+                const fingerprint = event.metadata.metadata?.fingerprint ?? "";
+                if (values.alias) {
+                  void createImageAlias(fingerprint, values.alias);
+                }
+                void queryClient.invalidateQueries({
+                  queryKey: [queryKeys.images, project],
+                });
+                close();
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:97]:
```suggestion
```
I think we don't need this any more, the close call in the finally block below should trigger earlier.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+            loading={formik.isSubmitting}
+            disabled={!formik.isValid}
+            onClick={() => void formik.submitForm()}
+          >
+            Upload image
+          </ActionButton>
+        </>
+      }
+    >
+      <Form onSubmit={formik.handleSubmit}>
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:154]: We should hide the form during upload. Condition would be `!uploadState` to hide it.

File: src/api/images.tsx [hunk 26]
```diff
@@ -107,3 +109,28 @@ export const createImage = (
       .catch(reject);
   });
 };
+
+export const createImageFromTarball = (
```
Comment (edlerd) [inline src/api/images.tsx:113]:
```suggestion
export const uploadImage = (
```
More concise and just as precise? Also I think we won't always have a tarball here.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+    toastNotify.success(<>Image {uploaded}.</>);
+  };
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      void formik.setFieldValue("fileList", e.target.files);
+    }
+  };
+
+  const getImageToTarballBody = (fileList: FileList): File | FormData => {
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:44]:
```suggestion
  const getImageUploadBody = (fileList: FileList): File | FormData => {
```
Tarball seems wrong, as this is also handling the case of root image plus metadata file, that can be gzip, without a tarball.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+
+      if (!formData.has("metadata") || !formData.has("rootfs.img")) {
+        throw new Error(
+          "Both metadata (tar.xz / gzip) and rootfs.img files are required",
+        );
+      }
+
+      return formData;
+    }
+    return fileList[0];
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:61]: Can we turn the condition around `If filelist.length === 1 { ... }` to improve two things:
1. we will have a positive comparison, making it easier to read and
2. most of the function will not be double indented.

**Review by edlerd — COMMENTED** [review]
Thanks for the adjustments, this is now much more straight forward to read.

File: src/api/images.tsx [hunk 29]
```diff
@@ -107,3 +109,28 @@ export const createImage = (
...
+  body: File | FormData,
+  alias: string,
+  setUploadState: (value: UploadState) => void,
+): Promise<LxdOperationResponse> => {
+  return new Promise((resolve, reject) => {
+    axios
+      .post(`/1.0/images`, body, {
+        headers: {
+          "X-LXD-public": alias,
```
Comment (mas-who) [inline src/api/images.tsx:122]: Why is `alias` assigned to the `X-LXD-public` header?

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 30]
```diff
@@ -0,0 +1,181 @@
...
+  };
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      void formik.setFieldValue("fileList", e.target.files);
+    }
+  };
+
+  const getImageToTarballBody = (fileList: FileList): File | FormData => {
+    if (fileList.length != 1) {
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:45]: should we add an extra condition here for `fileList.length > 2` and throw an error? Not sure about what copy to provide though, the thinking here is that we could prevent a potentially expensive API call (if the files are large).

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 31]
```diff
@@ -0,0 +1,181 @@
...
+    initialValues: {
+      alias: "",
+      isPublic: false,
+      fileList: null,
+    },
+    validationSchema: Yup.object().shape({
+      alias: Yup.string(),
+    }),
+    onSubmit: (values) => {
+      if (values.fileList) {
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:78]: I don't think we need this check here if we disable form submission for when `fileList` is `null`

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 32]
```diff
@@ -0,0 +1,181 @@
...
+  const getImageToTarballBody = (fileList: FileList): File | FormData => {
+    if (fileList.length != 1) {
+      // Sorting by Size. The metadata file is very likely to be smaller than the image itself.
+      const formData = new FormData();
+      const sortedFiles = Array.from(fileList).sort((a, b) => a.size - b.size);
+
+      formData.append("metadata", sortedFiles[0]);
+      formData.append("rootfs.img", sortedFiles[1]);
+
+      if (!formData.has("metadata") || !formData.has("rootfs.img")) {
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:53]: Is this check necessary? If we have more than 1 file, then surely both form data fields will be populated.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 33]
```diff
@@ -0,0 +1,181 @@
...
+      isPublic: false,
+      fileList: null,
+    },
+    validationSchema: Yup.object().shape({
+      alias: Yup.string(),
+    }),
+    onSubmit: (values) => {
+      if (values.fileList) {
+        createImageFromTarball(
+          getImageToTarballBody(values.fileList),
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:80]: have you checked that if an error was thrown from this function at this point, it does not crash the entire application? I don't think the `.catch` block will catch the error from this function since it's not part of `createImageFrromTarball`.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 34]
```diff
@@ -0,0 +1,181 @@
...
+          values.alias || "",
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:81]: I think this should be `values.isPublic`?

File: src/pages/images/actions/forms/UploadImageForm.tsx (hunk 25 above)
Comment (mas-who) [reply src/pages/images/actions/forms/UploadImageForm.tsx:154]: Yes, maybe use the same approach as the other components that have this upload pattern? We can hide the form with `className={uploadState ? "u-hide" : ""}`

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 35]
```diff
@@ -0,0 +1,181 @@
...
+            type="button"
+            onClick={close}
+          >
+            Cancel
+          </Button>
+          <ActionButton
+            appearance="positive"
+            className="u-no-margin--bottom"
+            loading={formik.isSubmitting}
+            disabled={!formik.isValid}
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:146]: I think we should disable submission if no files are selected
```suggestion
            disabled={!formik.isValid || !formik.values.fileList}
```

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 36]
```diff
@@ -0,0 +1,181 @@
...
+          error={formik.touched.alias ? formik.errors.alias : null}
+        />
+        <Input
+          {...formik.getFieldProps("isPublic")}
+          type="checkbox"
+          label="Make the image publicly available"
+          error={formik.touched.isPublic ? formik.errors.isPublic : null}
+        />
+        {/* hidden submit to enable enter key in inputs */}
+        <Input type="submit" hidden value="Hidden input" />
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:175]: This should also be disabled if no files are selected.

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 37]
```diff
@@ -0,0 +1,181 @@
...
+          type="file"
+          name="fileList"
+          label="Image backup file"
+          onChange={changeFile}
+          multiple
+        />
+        <Input
+          {...formik.getFieldProps("alias")}
+          type="text"
+          label="Alias"
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:165]: Let's add a placeholder for this input "Enter alias"

File: src/pages/images/actions/forms/UploadImageForm.tsx (hunk 25 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/UploadImageForm.tsx:154]: Wait, why would we hide the upload form when we have put the progress bar inside it?

File: src/pages/images/actions/forms/UploadImageForm.tsx (hunk 31 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/UploadImageForm.tsx:78]: image

Comment (mas-who) [reply src/pages/images/actions/forms/UploadImageForm.tsx:78]: aah I see, type checks

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 38]
```diff
@@ -0,0 +1,197 @@
...
+        formData.append("metadata", sortedFiles[0]);
+        formData.append("rootfs.img", sortedFiles[1]);
+
+        return formData;
+      }
+    } catch (e) {
+      close();
+      toastNotify.failure(`Image upload failed`, e);
+      throw e;
+    }
```
Comment (edlerd) [inline src/pages/images/actions/forms/UploadImageForm.tsx:68]: I think we shouldn't throw/catch in here, the check if have more than two files should be in line85 with the `if (values.fileList)` check.

**Review by edlerd — DISMISSED** [review]

File: src/pages/images/actions/forms/UploadImageForm.tsx (hunk 38 above)
Comment (Kxiru) [reply src/pages/images/actions/forms/UploadImageForm.tsx:68]: Thank you, please see my review :)

**Comment by mas-who** [comment]
@Kxiru I just tested uploading the files again and it's failing on unified tarballs (see below). I suspect it's because in your axios API call, `Content-Type` header is not set to `application/octet-stream`. Please would you look into this thanks.

Screenshot from 2024-08-22 14-53-06

**Comment by Kxiru** [comment]
> @Kxiru I just tested uploading the files again and it's failing on unified tarballs (see below). I suspect it's because in your axios API call, `Content-Type` header is not set to `application/octet-stream`. Please would you look into this thanks.
> 
> Screenshot from 2024-08-22 14-53-06
> 
> Edit: the same occurs for split image upload as well. Please check both cases.

Hmm... Looking into this as I am not having this issue locally, only on the demo server...

File: src/pages/images/actions/forms/UploadImageForm.tsx [hunk 39]
```diff
@@ -0,0 +1,196 @@
...
+      alias: "",
+      isPublic: false,
+      fileList: null,
+    },
+    validationSchema: Yup.object().shape({
+      alias: Yup.string(),
+    }),
+    onSubmit: (values) => {
+      if (values.fileList) {
+        if (values.fileList.length <= 2) {
```
Comment (mas-who) [inline src/pages/images/actions/forms/UploadImageForm.tsx:76]: I think we had a misunderstanding, the `else` block for `if (values.fileList)` should be kept since it would inform the user if they didn't select any files. Then one level down, we can do an early return like:
```
if (values.fileList.length > 2) {
        toastNotify.failure(...)
        return;
   }
```
so that we don't need to wrap the entire api call in an if else block.

**Review by edlerd — APPROVED** [review]
LGTM, thanks for adding the upload.
